File: go-katas/practice/matchingengine/order.go
package matchingengine

import "errors"

// Side is the direction of an order: buying or selling.
type Side int

// The two order sides.
const (
	// Buy orders want to acquire quantity; they match against the ask side and,
	// when resting, sort high→low (the highest bid is the best bid).
	Buy Side = iota
	// Sell orders want to dispose of quantity; they match against the bid side
	// and, when resting, sort low→high (the lowest ask is the best ask).
	Sell
)

// String returns the human-readable name of the side.
func (s Side) String() string {
	panic("TODO: implement")
}

// OrderType is how an order behaves when it cannot fully match: a Limit order
// rests, a Market order never does.
type OrderType int

// The two order types.
const (
	// Limit orders carry a worst acceptable price; any unfilled remainder rests
	// on the book at that price.
	Limit OrderType = iota
	// Market orders take whatever liquidity exists at any price; any unfilled
	// remainder is discarded (a market order never rests on the book).
	Market
)

// String returns the human-readable name of the order type.
func (t OrderType) String() string {
	panic("TODO: implement")
}

// Order is a single instruction to buy or sell Qty units. Price is in integer
// ticks (the exchange minimum increment) — money is never a float64. Price is
// ignored for Market orders.
type Order struct {
	ID    string
	Side  Side
	Type  OrderType
	Price int64
	Qty   int64
}

// Trade is one execution: TakerID is the incoming (aggressive) order, MakerID is
// the resting order it matched, Price is the maker's price, and Qty is the
// quantity that changed hands.
type Trade struct {
	MakerID string
	TakerID string
	Price   int64
	Qty     int64
}

// ErrUnknownOrder is returned when an operation references an order id that is
// not resting on the book.
var ErrUnknownOrder = errors.New("matchingengine: unknown order")

// ErrEngineClosed is returned by Engine.Submit after the engine has been closed.
var ErrEngineClosed = errors.New("matchingengine: engine is closed")

File: mm-katas/src/main/java/com/katas/k01_orderbook/Order.java
package com.katas.k01_orderbook;

/**
 * An order submitted to the book.
 *
 * <p>Prices are integer <em>ticks</em> — never use floating point for price comparison. For a
 * {@link OrderType#MARKET} order the {@link #price()} field is ignored.
 *
 * @param id      unique order id (used for cancellation and to tag fills)
 * @param account owning account (used for self-trade prevention)
 * @param side    buy or sell
 * @param type    limit or market
 * @param price   limit price in ticks (ignored for market orders)
 * @param qty     order quantity; must be positive
 */
public record Order(long id, String account, Side side, OrderType type, long price, long qty) {

    public Order {
        if (qty <= 0) {
            throw new IllegalArgumentException("qty must be positive: " + qty);
        }
    }

    /** A limit order resting at {@code price} ticks. */
    public static Order limit(long id, String account, Side side, long price, long qty) {
        return new Order(id, account, side, OrderType.LIMIT, price, qty);
    }

    /** A market order that executes against the best available prices. */
    public static Order market(long id, String account, Side side, long qty) {
        return new Order(id, account, side, OrderType.MARKET, 0L, qty);
    }
}
